### PR TITLE
Regularization

### PR DESCRIPTION
**Pull request by congxie1108**
*Description of changes:*

Implemented ZoneoutCell (both mxnet version), RNNZoneoutCell (the version fixes the issue that the output to the next layer and in the states using different masks), activation regularization, and temporal activation regularization.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/gluonts/model/deepar/_network.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 
 # Third-party imports
 import mxnet as mx
+import gluonnlp as nlp
```
Comment (szhengac) [inline src/gluonts/model/deepar/_network.py]: I don't see where we use gluonnlp in the file

File: src/gluonts/mx/rnn_extend/rnn_cell.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+__all__ = ['ZoneoutCell', 'RNNZoneoutCell', 'AccumulateStatesCell', "ActivationRegularizationLoss", "TemporalActivationRegularizationLoss"]
+
+# Third-party imports
+from mxnet import symbol, ndarray
+from mxnet.gluon.rnn import ModifierCell, BidirectionalCell, SequentialRNNCell
+from mxnet.gluon.rnn.rnn_cell import _format_sequence, _mask_sequence_variable_length, _get_begin_state
+from mxnet.gluon import tensor_types
+from mxnet.gluon.loss import Loss
+
+class ZoneoutCell(ModifierCell):
```
Comment (szhengac) [inline src/gluonts/mx/rnn_extend/rnn_cell.py:32]: Add docstring to each parameter

File: src/gluonts/mx/rnn_extend/rnn_cell.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+            new_states.append(next_states[0])
+        else:
+            new_states = [F.where(mask(p_states, new_s), new_s, old_s) for new_s, old_s in
+                                  zip(next_states, states)] if p_states != 0. else next_states
+
+        self._prev_output = output
+
+        return output, new_states
+
+class RNNZoneoutCell(ModifierCell):
```
Comment (szhengac) [inline src/gluonts/mx/rnn_extend/rnn_cell.py:195]: Why do we have two zoneout cells?

File: src/gluonts/model/deepar/_network.py (hunk 1 above)
Comment (congxie1108) [reply src/gluonts/model/deepar/_network.py]: You are right. My mistake.
I was trying to use AR and TAR from gluonnlp, but ended up with implementing my own version, and forgot to delete the import.

File: src/gluonts/model/deepar/_network.py [hunk 4]
```diff
@@ -425,6 +464,27 @@ def hybrid_forward(
         # need to mask possible nans and -inf
         loss = F.where(condition=loss_weights, x=loss, y=F.zeros_like(loss))
 
+        if self.alpha or self.beta:
+            # get accumulated outputs
+            outputs_raw = []
+            outputs_dropped = []
+            for state, state_info in zip(self.states, self.rnn.state_info()):
+                if 'state_index' in state_info:
```
Comment (szhengac) [inline src/gluonts/model/deepar/_network.py]: so the format is [dropped_h, c, h]?

File: src/gluonts/mx/rnn_extend/rnn_cell.py (hunk 3 above)
Comment (congxie1108) [reply src/gluonts/mx/rnn_extend/rnn_cell.py:195]: Zoneout is functioning the same as the original one in mxnet, which uses different masks for output and the output in states
RNNZoneout uses the same mask for the output and the output inside states (the first element of states)
And for both of them, I add a flag "preserve_raw_output" to indicate whether to append the output without dropout to the states, that's why I also need to implement not only RNNZoneout, but also the original Zoneout.

File: src/gluonts/model/deepar/_network.py (hunk 4 above)
Comment (congxie1108) [reply src/gluonts/model/deepar/_network.py]: for each rnn layer, after accumulation, the final version of states will be [dropped_h, c, raw_h, [list of dropped_h], [list of raw_h]]
since there are 2 layers, the states of them will be concatenated.
so there should be totally 10 elements in the final states.

Comment (congxie1108) [reply src/gluonts/model/deepar/_network.py]: the logic is:
for raw lstm, the states if [dropped_h, c]
after add zoneout with the option "preserve_raw_output=True", it will append raw_h to the states, then the states become [dropped_h, c, raw_h]
then, after add AccumulateStatesCell with option "index_list=[0, -1]", it will accumulate dropped_h and raw_h throughout timestamps and append them to states, so finally the states become [dropped_h, c, raw_h, [list of dropped_h], [list of raw_h]]

File: src/gluonts/mx/rnn_extend/rnn_cell.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+        if prev_output is None:
+            prev_output = F.zeros_like(next_output)
+
+        output = (F.where(mask(p_outputs, next_output), next_output, prev_output)
+                  if p_outputs != 0. else next_output)
+        
+        if self.preserve_raw_output:
+            # the output is stored in the last element of states, thus skip it
+            new_states = [F.where(mask(p_states, new_s), new_s, old_s) for new_s, old_s in
+                                  zip(next_states, states[0:-1])] if p_states != 0. else next_states
```
Comment (szhengac) [inline src/gluonts/mx/rnn_extend/rnn_cell.py:93]: can remove 0, [:-1] is enough

File: src/gluonts/mx/rnn_extend/rnn_cell.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+
+class AccumulateStatesCell(ModifierCell):
+    """
+    Accumulate part of the states in base cell.
+    To make it work, AccumulateStatesCell must be the last ModifierCell when stacking
+    """
+    def __init__(self, base_cell, index_list):
+        super(AccumulateStatesCell, self).__init__(base_cell)
+        self.index_list = index_list
```
Comment (szhengac) [inline src/gluonts/mx/rnn_extend/rnn_cell.py:281]: what is index list?

Comment (congxie1108) [reply src/gluonts/mx/rnn_extend/rnn_cell.py:281]: to indicate which elements in states needs to be accumulated

File: src/gluonts/mx/rnn_extend/rnn_cell.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+        begin = self.base_cell.begin_state(func=func, **kwargs)
+        for index in self.index_list:
+            # placeholder for the accumulated states
+            begin.append([])
+        self.base_cell._modified = True
+        return begin
+
+    def hybrid_forward(self, F, inputs, states):
+        # identity mapping
+        next_output, next_states = self.base_cell(inputs, states)
```
Comment (szhengac) [inline src/gluonts/mx/rnn_extend/rnn_cell.py:304]: states here has 3 elements right?

File: src/gluonts/mx/rnn_extend/rnn_cell.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+        # append accumulated states
+        accumulated_states = []
+        for i, index in enumerate(self.index_list):
+            accumulated_states.append(states[len(states)-len(self.index_list)+i] + [next_states[index]])
+        next_states.extend(accumulated_states)
+
+        return next_output, next_states
```
Comment (szhengac) [inline src/gluonts/mx/rnn_extend/rnn_cell.py:311]: the next_states should be passed to the next time stamp, but it is accumulated?

File: src/gluonts/mx/rnn_extend/rnn_cell.py (hunk 7 above)
Comment (congxie1108) [reply src/gluonts/mx/rnn_extend/rnn_cell.py:304]: yes, next_states should only have 3 elements

Comment (congxie1108) [reply src/gluonts/mx/rnn_extend/rnn_cell.py:304]: states have 5 elements

File: src/gluonts/model/deepar/_network.py (hunk 4 above)
Comment (szhengac) [reply src/gluonts/model/deepar/_network.py]: 10 elements?

File: src/gluonts/mx/rnn_extend/rnn_cell.py (hunk 8 above)
Comment (congxie1108) [reply src/gluonts/mx/rnn_extend/rnn_cell.py:311]: yes, when execute "next_output, next_states = self.base_cell(inputs, states)", although there are 5 elements in states, the base cell will ignore the elements on the tail of states, and only use the first 2 elements of states, then return the next_states of [dropped_h, c, raw_h]

File: src/gluonts/mx/rnn_extend/rnn_cell.py (hunk 7 above)
Comment (szhengac) [reply src/gluonts/mx/rnn_extend/rnn_cell.py:304]: btw, the base_cell accepts number of elements more than the predefined? for example, lstm only has 2 states, but we input 5 elements to it.

File: src/gluonts/mx/rnn_extend/rnn_cell.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+        self.base_cell._modified = True
+        return begin
+
+    def hybrid_forward(self, F, inputs, states):
+        # identity mapping
+        next_output, next_states = self.base_cell(inputs, states)
+        # append accumulated states
+        accumulated_states = []
+        for i, index in enumerate(self.index_list):
+            accumulated_states.append(states[len(states)-len(self.index_list)+i] + [next_states[index]])
```
Comment (szhengac) [inline src/gluonts/mx/rnn_extend/rnn_cell.py:308]: it seems that the first part can be reduced to states[i - len(self.index_list)].

File: src/gluonts/mx/rnn_extend/rnn_cell.py (hunk 7 above)
Comment (congxie1108) [reply src/gluonts/mx/rnn_extend/rnn_cell.py:304]: states have 5 elements, and next_states have 3 elements at first, and then have 5 elements when it is returned.
I've confirmed it by adding some assert to debug.

File: src/gluonts/model/deepar/_network.py (hunk 4 above)
Comment (congxie1108) [reply src/gluonts/model/deepar/_network.py]: deepar has 2 rnn layers by default
every rnn layer has 5 elements, concatenating 2 layers gets 10.
